Allows blacklisting words from IC chat via config

ike709 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

~~Blocks netspeak in IC using regex.~~ Implements a config file for filtering IC, for the purpose of blocking netspeak. It's still allowed in asay/dsay/ooc/etc. I'm aware that this list is not exhaustive and that it can be fairly easily bypassed. But this blocks most of the common occurrences so it'll make life easier for admins, and it'll be obvious when people are intentionally breaking the rule by bypassing it.

It also doesn't block words that contain blacklisted words in them. For example, "lol" (if configured) would be blocked but "lollygagging" would not.

I expect this to cause at least three (3) angry rants on reddit.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Netspeak in IC is against the rules. MUH IMMERSHUN, and all that.


#### EDIT: I'm too lazy to rewrite the whole PR desc, but this is now a generic IC blacklist that is edited via config. It's no longer specifically targeting netspeak. It is not blocking anything by default.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: ike709
add: IC chat can now be filtered via config to prevent things like netspeak or anything else against the rules.
/:cl:

Wilchenx commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	if(!message || message == "")
 		return
 
+	var/static/regex/netspeak = regex("\\b(lol|brb|wtf|idk|omg|lmao|btw|ikr|rofl|tbh|nvm|thx|afaik)\\b", "i")
```
add "reddit"

ike709 replied:
Don't tempt me.

Tlaltecuhtli replied:
add "owo" and "uwu" as they are one of the major crimes of this kind

optimumtact replied:
please don't use code comments unless you are pointing out an actual code issue or are a maintainer please.

They're otherwise wildly disruptive to the rest of the conversation.

Fikou commented:
what about gibbering messages

Wilchenx commented:
A valid concern is clown name policy, if the player was to use one of the blacklisted words as their name, noone would be able to call them out on the radio.

redmoogle reviewed (changes requested):
words to also include
'netspeak'
'ooc'
'ic'
'ROF(T)L'
'Rly'
'pls'
'ez'
'cuz'
not all but should help alot

nicbn commented:
Well it's not that rare to see (hear) someone saying something like lol irl, that's really dumb

redmoogle commented:
why not ban the world netspeak itself

> Well it's not that rare to see (hear) someone saying something like lol irl, that's really dumb

yes but it's frowned upon in-game because you should say haha instead or something like that

FloranOtten commented:
lol and lmao are words used IRL  as well. You may call it cringe but there you go. I think I hear the word lmao spoken more than I hear the entire sentence spoken.

HomocidlePancakes commented:
>lol and lmao are words used IRL as well. You may call it cringe but there you go. I think I hear the word lmao spoken more than I hear the entire sentence spoken.

Sorry. You have to be atleast 18 to play on TG. How did you end up in this unwholesome hellhole child?

partykp commented:
What if you made all the instances of "laughing" words like lol and lmao just make the offender say *laugh

ike709 committed:
Blocks netspeak in IC

nervere commented:
Is it possible for you to make what words are filtered a config option?
What about a config option with the ability to use regex of our own?
Either works, I just feel this being hardcoded is not a good idea.

OverlyGeneric commented:
I don't believe this will cause much trouble on Reddit. they may actually commend you for relieving their frustrations.

good work.

nervere commented:
> I don't believe this will cause much trouble on Reddit. they may actually commend you for relieving their frustrations.
> 
> good work.

No one cares about Reddit.

JJRcop commented:
This should be a config option in the interest of downstreams

optimumtact commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	if(!message || message == "")
 		return
 
+	var/static/regex/netspeak = regex("\\b(lol|brb|wtf|idk|omg|lmao|btw|ikr|rofl|tbh|nvm|thx|afaik|ooc|ic|pls|plz)\\b", "i")
```
does this actually ensure the regex is only compiled once?

Also I suspect a chain of findtext calls will actually be faster.

third this should be configurable list in config

@MrStonedOne are you happy for a wordfilter to be applied?

MrStonedOne replied:
Make it a configurable list and kick the bucket down to the headmins, but this would need to be changed as static vars are inited with global vars before anything else happens, including config loading.

ike709 replied:
I don't know how to check if the regex is only compiled once.

But I'll profile this vs findtext and rework it into a config when I have time, if you are willing to merge this once those improvements have been made. Otherwise I won't waste my time.

optimumtact replied:
Im not opposed to it

make sure you add a config option to globally enable/disable it as well, or just use the empty state of the list to skip any checks

ike709 replied:
Turned this into a config.

ike709 committed:
Allows gibbering people to still use netspeak

One of the brain damage lines would've been blocked.
Also adds a few more instances of netspeak.

4dplanner commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
+	var/static/regex/netspeak = regex("\\b(lol|brb|wtf|idk|omg|lmao|btw|ikr|rofl|tbh|nvm|thx|afaik|ooc|ic|pls|plz)\\b", "i")
+	if(!derpspeech && findtext(message, netspeak))
+		to_chat(src, "<span class='warning'>Netspeak is not permitted in character!</span>")
+		return
```
should allow you to force it with a warning imo

optimumtact replied:
why would we willingly let people say something that can net them a 5 minute ban?

4dplanner replied:
There are legitimate cases, like saying "my nephew posted `lol i rekt a scrub` on facebook!", in which you're referring not using

optimumtact replied:
also stop using code comments on things that aren't not directly technical feedback

Wilchenx commented:
>this being hardcoded is not a good idea.

Absolutely correct, downstreams may not want a fixed list.

ike709 commented:
I'll rework this into a config later today when I have free time.

bgobandit commented:
you are absolutely certain that this blocks only the words themselves and not words containing those character strings, correct?

MrStonedOne commented:
The `\\b` matches a space

JJRcop commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 		return
 
+	var/static/regex/netspeak = regex("\\b(lol|brb|wtf|idk|omg|lmao|btw|ikr|rofl|tbh|nvm|thx|afaik|ooc|ic|pls|plz)\\b", "i")
+	if(!derpspeech && findtext(message, netspeak))
```
```suggestion
	if(!forced && findtext(message, netspeak))
```

ike709 commented:
It's a config file now.

JJRcop commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,12 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 
+	if(GLOB.in_character_filter.len)
+		var/regex/filter = regex("\\b([jointext(GLOB.in_character_filter, "|")])\\b", "i")
+		if(!derpspeech && findtext(message, filter))
```
```suggestion
		if(!forced && findtext(message, filter))
```

ike709 replied:
Done.

ike709 commented:
Unless any other requested changes arise, I think this is ready.

optimumtact commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,12 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 		return
 
+	if(GLOB.in_character_filter.len && !forced)
+		var/regex/filter = regex("\\b([jointext(GLOB.in_character_filter, "|")])\\b", "i")
```
you need to compile this once at world init please

ike709 replied:
Done.

ike709 committed:
Config

ike709 committed:
derpspeech -> forced, and check it sooner

ike709 committed:
Satisfies the fruit

nervere commented on code/modules/mob/living/say.dm:
```diff
@@ -93,6 +93,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 
+	if(GLOB.in_character_filter.len && !forced)
+		if(findtext(message, config.ic_filter_regex))
+			to_chat(src, "<span class='warning'>That message contained a word prohibited in IC chat!</span>")
```
this should probably be a configurable message as well

ike709 replied:
People can already click the configurable rules button at the top of the game window, but I will alter the message to mention it.

ike709 committed:
Da Rulez

YakumoChen commented:
isn't this more of an administrative issue? why do we need to hardcode this

FloranOtten commented:
I am also not a fan of having hardcoded filters. We don't need god damn ban bots. Who exactly was so bothered by someone saying 'lol'? Fuck, I've heard people say that in common speech.

Alek2ander commented:
The filter shouldn't apply to the binary channel because it's literally netspeak

optimumtact commented:
@YakumoChen apparently you're so stuck in your bunker you haven't seen the drama going on right now

Mark9013100 commented:
lol

dr-bee commented:
Wouldnt it be a nicer solution to have LOL just trigger the laughing emote and not the word lol instead of blocking the entire message?

FloranOtten commented:
> @YakumoChen apparently you're so stuck in your bunker you haven't seen the drama going on right now

Drama?

Tlaltecuhtli commented:
it might break immersion more as for stuff like force speech nanites on "lol" and having everyone spammed by unknown ooc in ic warns

this change is dumb let the banbots handle their dopamine level by giving out 5 min bans

CPTANT commented:
wtf and omg are just the written abbreviation of what the fuck and oh my God . People say that all the time. 

And lol is used all the time here.

ike709 commented:
Imagine unironically saying "lol" as an adult.

CPTANT commented:
Well "lol" has been the Dutch word for fun since a couple of centuries it's not that weird.

Wilchenx commented:
We're a english speaking community

Yttriums commented:
Is netspeak more immersion breaking than having players get popups when they say specific disallowed words?

Wilchenx commented:
Lifeweb weekbans you for using netspeak

So i'd say yes, its frowned upon to use netspeak while roleplaying.

CPTANT commented:
We're a space station 13 community

Wilchenx commented:
Ale jakie ma to znaczenie?

4dplanner commented:
> it might break immersion more as for stuff like force speech nanites on "lol" and having everyone spammed by unknown ooc in ic warns
> 
> this change is dumb let the banbots handle their dopamine level by giving out 5 min bans

that's why it checks the forced variable

JJRcop commented:
The title of the PR should be changed since recent additions to it have modified what it does, and the title isn't accurate anymore.

After recent changes, the PR adds the ability for server administrators to blacklist words, but does not come preloaded with any words anymore.

chaoko99 commented:
"Cuz" is something that isn't netspeak... It's short for "because", and real people have been saying it for ages.

Yttriums commented:
The game takes place in a future with advanced electronic communication, obviously electronic communication will have influenced the way people speak just like it did in our world.

Netspeak IC is way more plausible than, say, having a clown on a corporate spaceship for entertaining adults in 2500ad.

Wilchenx commented:
surely noone will speak about reddit in 2500ad

ike709 commented:
@JJRcop Good point.

ike709 commented:
@CPTANT Well then all of the Dutch servers can just not put "lol" in the config file.

ike709 commented:
@chaoko99 Then I guess it's good that "cuz" was never blacklisted at any point even when it was hardcoded.

nervere commented:
If you cannot talk without using netspeak, I'm afraid this roleplaying game is not for you.

optimumtact commented:
this is a configurable thing if you want to argue about netspeak on tg please go to the forum policy discussion
